Add support for a {% debug %} extension tag.

ShaheedHaque opened the pull request:
This dumps the available variables, filters and tests. See Issue #174.
This is roughly equivalent to the Django Template Language {% debug %} tag, and typical usage like this:

`
        {% debug %}
`

produces output like this:

```
        {'context': {'_': <function _gettext_alias at 0x7f9ceabde488>,
                 'csrf_token': <SimpleLazyObject: 'lfPE7al...q3bykS4txKfb3'>,
                 'cycler': <class 'jinja2.utils.Cycler'>,
                 ...
                 'view': <polls.views_auth.Login object at 0x7f9cea2cbe48>},
        'filters': ['abs', 'add', 'addslashes', 'attr', 'batch', 'bootstrap',
                 'bootstrap_classes', 'bootstrap_horizontal',
                 'bootstrap_inline', ... 'yesno'],
        'tests': ['callable', 'checkbox_field', 'defined', 'divisibleby',
               'escaped', 'even', 'iterable', 'lower', 'mapping',
               'multiple_checkbox_field', ... 'string', 'undefined', 'upper']}
```

The output under Python2 and Python3 below 3.4 is slightly less compact, as pprint.format() did not introduce the 'compact' argument before 3.4.

davidism commented:
It would be interesting to support syntax like `{% debug html-table %}` to produce a table when you know you're in an HTML template.

ShaheedHaque commented:
To be honest, that seems like overkill given that this is a debug facility intended to efficiently serve the all-too-common scenario of "Hmmm, I'm not quite sure what all the context variables are" using built-in Python string formatting. If table support was needed, perhaps it would be better to introduce some kind of a magic {{ debug }} which returned the equivalent dict, and allowed the user to select what was needed and how it should be styled?

davidism commented:
That's the other thing I was thinking. Using a tag means you can't access the data directly, unless you were to turn it into a block tag like `call`.

ShaheedHaque commented:
True, but TBH, that sounds like a whole different thing: its not a tag for the reason you suggest, nor is the idea of a "magic" context variable that contains all other context variables/filters/tests a "nice" thing. That all but calls for new syntax {{ %reserved-thing% }}, for example. I believe the requirement being addressed here is much simpler than full scale access to that data.

ShaheedHaque commented:
If you'd prefer not to accept this PR, please do let me know (the downstream project I'd originally submitted this to might be an alternative home).

davidism commented:
I'd like to hear from @mitsuhiko or @ThiefMaster before merging.

In order to use it, you need to know about and enable the extension first, add a tag, and you only get a squashed, non-interactive representation that isn't JSON compatible.

There are already things like [Flask-DebugToolbar](https://github.com/mgood/flask-debugtoolbar). The user can also set a breakpoint in a debugger and look at the context directly:

![pycharm-jinja-debug](https://user-images.githubusercontent.com/1242887/35107774-a2802952-fc26-11e7-85ea-fee19715c170.png)

davidism commented on jinja2/ext.py:
```diff
@@ -434,6 +446,57 @@ class AutoEscapeExtension(Extension):
 
+class DebugExtension(Extension):
+    """
+    A {% debug %} tag that dumps the available variables, filters and tests.
```
Use inline code markup

    ``{% debug %}``

ShaheedHaque replied:
Ack.

davidism commented on jinja2/ext.py:
```diff
@@ -434,6 +446,57 @@ class AutoEscapeExtension(Extension):
+class DebugExtension(Extension):
+    """
+    A {% debug %} tag that dumps the available variables, filters and tests.
+    This is roughly equivalent to the Django Template Language {% debug %} tag,
```
No need to mention Django here, this won't mean anything if you haven't used the Django debug tag, and it's not really equivalent anyway.

ShaheedHaque replied:
Ack.

davidism commented on jinja2/ext.py:
```diff
@@ -434,6 +446,57 @@ class AutoEscapeExtension(Extension):
+    This is roughly equivalent to the Django Template Language {% debug %} tag,
+    and typical usage like this:
+
+        <pre>{% debug %}</pre>
```
Code blocks must be preceded with two colons: `::` or (preferably with non-Python) `.. codeblock:: html+jinja`.

ShaheedHaque replied:
I had not appreciated the code was being processed for markup. I will look into this.

davidism replied:
Pygments recognizes `html+jinja`, use that over `html`.

davidism commented on jinja2/ext.py:
```diff
@@ -434,6 +446,57 @@ class AutoEscapeExtension(Extension):
+        # We set the depth since the intent is basically to show the top few
+        # names. TODO: provide user control over this?
+        #
+        if (version_info[0], version_info[1]) >= (3,4):
```
`version_info[:2]`

ShaheedHaque replied:
Duh.

davidism commented on tests/test_ext.py:
```diff
@@ -240,6 +240,22 @@ class T2(Extension):
+        env = Environment(extensions=['jinja2.ext.debug'])
+        tmpl = env.from_string('''Hello{% debug %}Bye''')
+        out = tmpl.render()
+        out = out.replace('&#x27;', "'").replace('&lt;', '<').replace('&gt;', '>')
```
Why replace these?

ShaheedHaque replied:
If an assert fails, the debug output is much easier to read, and so the tests are easier to maintain.

ShaheedHaque commented:
Of course, it is completely reasonable that you and the community must feel comfortable with what is proposed. If its not the right thing for Jinja, so be it. In the meantime, thanks for the excellent feedback, I will address the points noted.

On the general comments...

- about this being an extension. Yes, I packaged it that way because I thought that seemed like a plausible approach. I'm fine if you prefer to make it part of the default tag set.
- about the output not being JSON, I should just point out that the data exposed is commonly not vanilla data, at least in the Django context. There *are* ellipses present, but that's only because I specifically limited the output depth. If you prefer, I could remove that limit. Barring that, it looks like JSON to me...though I've not reviewed how pprint works..is there some case you have in mind?
- about the output not being interactive. I'm not sure quite what you had in mind, but interaction was not something I think belongs in this context. 
- about alternate debug facilities. I use PyCharm too, but it turns out that at least for Django-users, breakpoints are not supported with Jinja...only DTL is [supported](https://www.jetbrains.com/help/pycharm/debugging-django-templates.html). There are also plenty of scenarios where I think a simple dump is useful.

ShaheedHaque commented on jinja2/ext.py:
```diff
@@ -4,13 +4,24 @@
+    # Python3
+    from html import escape
+except ImportError:
+    from cgi import escape
```
Thanks for the tip. I'll look into it.

davidism commented:
JSON uses double quotes around strings. Values that aren't JSON types should be converted to string (this is not trivial).

Interactive as in I can access it as a varaible and render it how I want.

PyCharm says it supports debugging Jinja templates, so if that isn't the case it needs to be reported to PyCharm. Are you sure you selected "Jinja2" for the template language in the project? Are you sure that the file is detected as a Jinja template (it can be changed in the file association settings)?

ShaheedHaque commented:
I think I've addressed the comments, but it is not clear to me how to verify the markup-related suggestions (I had to work around an unrelated doc build failure locally, but the Travis build fails in some way that seems likely related to my change...I'll have to experiment to fix it).

- On the JSON format. I'm happy to look at that if that becomes the blocking issue.
- On the interactive thing. I have no idea how to implement that and it is not of interest to me.
- On the PyCharm support. The link I provided is to the official PyCharm docs: the combination of Django and Jinja is not expected to work.

davidism commented on jinja2/ext.py:
```diff
@@ -434,6 +442,59 @@ class AutoEscapeExtension(Extension):
+
+    produces output like this:
+
+    ``
```
Use `::` for Python code blocks. Double backticks are only for inline code.

ShaheedHaque replied:
OK. Can you educate me on where to look to see what is produced? I've tried running sphinx by hand, and searched in vain for where this stuff ends up under the docs/_build tree. (I'm a sphinx newb).

ShaheedHaque commented:
I've no idea why the doc build is failing under Travis:

> Warning, treated as error:
/home/travis/build/pallets/jinja/docs/templates.rst:1314:Could not lex literal_block as "jinja". Highlighting skipped.
ERROR: InvocationError: '/home/travis/build/pallets/jinja/.tox/docs-html/bin/sphinx-build -W -b html -d /home/travis/build/pallets/jinja/.tox/docs-html/tmp/doctrees docs docs/_build/html'

I *do* get a different error locally, but if I make this change on about line 307 of docs/extensions.rst:

> \-   \:members\: push, look, eos, skip, \_\_next\_\_, next_if, skip_if, expect
> \+   \:members\: push, look, eos, skip, next_if, skip_if, expect

then I get a clean local build. I don't see how this can be related to my changes, so I fear I'm a bit stuck.

ShaheedHaque commented:
I note that PR #802 has the same doc build failure.

dwt commented:
@maintainers: any progress on this? I've just been searching for this again too.

davidism commented:
Continued and merged in #983.
